src/wasm/wasm-type.cpp

```diff
@@ -3968,6 +3968,188 @@ TypeBuilder::BuildResult TypeBuilder::build() {
   return {state.results};
 }
 
+TypeBuilder::BuildResult TypeBuilder::groupAndBuild() {
+  assert(impl->recGroups.size() == 0 &&
+         "Cannot use groupAndBuild with explicitly set recursion groups");
+
+  // Recursion groups and type ordering only matter for isorecursive types.
+  if (getTypeSystem() != TypeSystem::Isorecursive) {
+    return build();
+  }
+
+  const size_t numTypes = size();
+
+  // Use Tarjan's strongly connected components algorithm
+  // (https://en.wikipedia.org/wiki/Tarjan%27s_strongly_connected_components_algorithm)
+  // to organize the types into topologically sorted recursion groups.
+
+  // Map HeapTypes in the builder to indices.
+  std::unordered_map<HeapType, size_t> typeIndices;
+  for (size_t i = 0; i < numTypes; ++i) {
+    typeIndices.insert({impl->entries[i].get(), i});
+  }
+
+  // The metadata used by Tarjan's SCC algorithm.
+  struct Node {
+    std::vector<size_t> succIndices;
+    size_t index = -1u;
+    size_t lowlink = -1u;
+    bool onStack = false;
+  };
+
+  // Allocate metadata for each HeapType in the builder.
+  std::vector<Node> nodes(numTypes);
+
+  // Populate the successor indices for each node by finding the predecessors of
+  // each HeapType. Since we are incrementing the successor index that gets
+  // recorded, we can avoid recording duplicate successors by checking just the
+  // last recorded successor for each predecessor.
+  for (size_t succIndex = 0; succIndex < numTypes; ++succIndex) {
+    HeapType succ = impl->entries[succIndex].get();
+    for (auto pred : succ.getReferencedHeapTypes()) {
+      auto predIndexIt = typeIndices.find(pred);
+      if (predIndexIt == typeIndices.end()) {
+        // This is a basic or canonical child, so it is a predecessor of all the
+        // types we are building and won't affect their ordering or grouping.
+        continue;
+      }
+      // Record the successor if we have not already recorded it.
+      auto& succIndices = nodes[predIndexIt->second].succIndices;
+      if (succIndices.empty() || succIndices.back() != succIndex) {
```

Comment:
Why is it enough to check `back()` here and not look through the entire vector?

Reply:
We are recording the successor indices in order from zero to `numTypes`, so the highest successor index is always at the end of the vector and is never greater than the current successor index. That means that if there would be a repeat, it would be at the end of the vector.

Reply:
I see, thanks!